### DISCUSSION TITLE
Take pre-orders into account

### DISCUSSION THREAD

**Discussion by KodeStar** [joonaspaakko/audible-library-extractor#96] [Ideas]
Not sure if this is feasible with the data you can get, but for example:

<img width="314" alt="image" src="https://user-images.githubusercontent.com/5698566/169491868-0fd46418-c515-4141-9885-77084d96b5c7.png">

<img width="1049" alt="image" src="https://user-images.githubusercontent.com/5698566/169492025-c8887dff-fc87-459b-9e13-46b49ffd363d.png">

**Comment by joonaspaakko** [comment]
It is possible to get this info, but it's not as straight forward as I thought. Before I started thinking about this just now, I was considering it for the next version of the extension as a quick addition, but now I'm not sure if I'll ever add it. I wouldn't throw the idea out the window yet, but it does seem like a lot of work for a short term thing, given that pre-orders end up in your library eventually anyways.

> I did kinda mention it below but to be clear, I might have a good reason to fetch data from purchase history to get the `added date`, which would allow better filtering using that date. Technically you can currently filter by "added", but only in terms of the order / number of books. Like for example [first 6 books](https://joonaspaakko.github.io/my-audible-library/#/library?filterExtras=added%253A1-6) or any range, but fetching the added date from purchase history would allow to filter books by years or a date range. And that to me seems like a pretty enticing feature. If that is added, it would take next to nothing to check all the preorders and save that data. So it's still very much a possibility that it could happen.
>
> The added date used to be in the library page and also on the store page, but they took it away in like summer 2020.

- - - - - 

**This is just me thinking out loud...**

It might require scraping the purchase history, which isn't currently being done. Unless the pre-order status is visible on the series page or something (which would have its own issues). This screenshot is clearly from the store page, but just to check the status there would require essentially loading every single store page for every single book in a series, which isn't currently necessary and would require a ton of "extra" page loads... At which point scraping the entire purchase history  is likely faster. There has been the idea to fetch added dates from the purchase history, in which case it could be done at the same time. It would take some doing thought.

Another hurdle would be books that are not part of a series... Or part of a series you don't have any previous books in. For a book in a series you could simply fetch the status and display it in this "my books in the series" list, but if the book is not in a series there's no way to really display it, because it wouldn't be any kind of part of your library yet. You could make the argument that it's going to be in your library so you might as well show it in the library just like any book, which would require fetching all the data from the store page and series pages.

There's also the fact that pre-orders eventually end up in your library unless of course you cancel. So should I do all that work that is needed to fetch pre-orders or should the user to just wait a month or two and scan the library when it's released. I think it might fall outside of scope of the extension by a technicality if I wanted to be lazy about it: "Audible *Library* Extractor". And any extra scraping also does mean more waiting in the extraction process. If I was able to get this info from the series page or someplace that is already being scraped, compromise could be to only add it for existing series and show it in the "my books in the series" list.

**Comment by KodeStar** [comment]
No worries, I was just throwing out ideas. That said, if it was possible to mark books as “owned” as in one of the other ideas I posted, that might be a workaround and easier to implement as you already have a list of the other books in the series.

**Reply by joonaspaakko** [reply]
Yea. I'll get back to you about that.

**Reply by joonaspaakko** [reply]
@KodeStar, I just wanted to let you know about my thoughts on your 3 suggestions briefly because I think it will be a while before I can take a proper look at them. 

> All in a single comment because I'm lazy...

1. **Show pre-orders**
   - I've since decided that I will be extracting the purchase history to fetch the purchase date and pre-purchases can ride along with that.
   - I think they will be displayed just like any normal book in the library, with the added "I'm a pre-order" data point.
   - All this assuming there aren't any issues with the execution.
2. **Show missing books in the series subpage**, just like they are in the "my books in the series" list.
   - I've actually had this in my maybe pile since before the "my books in the series" list had missing books. This was the idea in terms of looks, but may e they could also have like a red border too or something, because gray covers do exist. Well maybe the muted numbers would be enough (or lower opacity/contrast). [The original idea](https://www.dropbox.com/s/c7xwoif95h69j44/ALE-not-in-library-series-list-2.png?raw=1).
   - This will require extracting "cover url" for every book in a series, but it's no longer a big deal since every series page is being fetched already, so I might as well add them.
3. **Manually adding books**
   - This might require some soul searching before I can say yay or nay. I'm starting to warm up to the idea of allowing users to add their own data into the mix. I think a few people suggested the ability to add notes to individual books.
   - I've said no so far because I feel like it's risky. Technically it's not more risky than storing important files on a single hard drive, but it just has not felt like a good idea because currently all data likes in Audible. Any custom data would have to be stored locally in the browser, so if you add a 1000 books or kotes or whatever and you remove the extension or your computer dies it's goodbye to hours of work. I would not add this without the ability to save the data locally either similar to the raw data import/export or with it... But still... I can't help but feel like it's a risk. 
   - But I mean if the user is made aware that they are the gatekeepers of any custom data they enter, then maybe it wouldn't hurt...
    - It's been a long time since I checked, but I believe the extension data sync max size is too small for that data, because it's not meant as a data storage. The local data as no limits.

**Reply by Britastic365** [reply]
I would LOVE the ability to add books manually.

Right now I have the biggest wishlist...because it's the only way to add books. I think I'm at about 7,000 audiobooks. I get the whole "browser stored data" problem, but what about being able to add it to our own github repo?
